src/diagonal.works/b6/gomod2nix.toml

```diff
@@ -0,0 +1,99 @@
+schema = 3
```

Comment:
This file is generated by the `gomod2nix` program; and is just a way for Nix to understand all the go dependencies.

Reply:
Good comments and thoughts. Let me say a few things. Forgive me for all the detail.

### Re: running dev builds in docker.

Indeed, it does work, but as you mention it's slow; too slow, I think, for meaningful development to be done. For example, I don't want to be using that to do JS development, or Python development. I _could_ be convinced for Go development, but again I miss anything incremental.

It's worth making a distinction, I think, between "incremental" builds/development and a kind of "full"/"release" build.

### Incremental vs "Full" builds

Let me go into a bit of detail; forgive me if any of this is already clear to you.

In the *devShell*, Nix provides all the necessary tooling to do _incremental_ builds. I.e. the go binaries are installed (and, in fact the python library is built and available). Inside the shell, we make full use of the Makefile, even leaving it to use the binaries in `./bin/...` that are built by the Makefile. This is necessary and useful; no-one wants to do "full" rebuilds for every development change.

Outside of the devShell, the flake provides _packages_. These are extremely useful for _running_ the tooling. For example, right now (supposing you've installed Nix with flakes enabled) you can run:

```
nix run github:silky/diagonal-b6/add-nix -- --help
```

and it runs the `b6` executable. 

Similarly, you can depend on the Python library with a very simple flake.nix; see - https://github.com/silky/example-b6-analysis/blob/main/flake.nix. And likewise you can build and run the docker image simply:

```
nix build github:silky/diagonal-b6/add-nix#b6-image
./result | docker load
```

These are both examples of "outside" tasks you might like to do.

Now, one could argue that in an ideal world none of this "outside" tooling would be necessary because, perhaps, we would have:

- Always be deploying nightly build of Python, say, to pypi or something,
- Always be publishing nightly docker images of the b6 tooling,
- And maybe somehow publish the Go binaries so they can be dependend on directly, if someone wishes.

And do all of these things in a way that ensures all the dependencies are always present (not the case, given gdal, IIRC).

Even if we _could_ do that, it doesn't completely help "outside" requirements for "dev" version of b6; i.e. if someone depends on a particular feature in some branch (maybe doesn't happen now, I don't know, but might be relevant at some point).

#### Sub-detail: How to build docker images?

Regarding docker, I am open-minded on building it "manually" (via the Dockerfile you construct) or via Nix, as I've done here: https://github.com/diagonalworks/diagonal-b6/pull/295/files#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R216.

To my eyes, the Nix version is cleaner, but of course I'm biased :)

The natural advantage to Nix for docker building is that it is _much_ faster to build when you make small changes; you don't need to wait for the entire `apt-get ...` round-trips.

I also don't quite like that it's duplicated, but as you mention, I don't really want to _force_ people to use Nix; so I wanted to have both options. Note that in any case, the docker image I build here is different from yours:

```
> docker image ls | grep b6
b6                                            latest    ec3003eb9932   25 hours ago    64.6MB
europe-docker.pkg.dev/diagonal-public/b6/b6   latest    21f0c4c584f9   13 months ago   348MB
```

I think the main reason is I _only_ place the `b6` binary on there; none of the others (and I don't include any v1 JS artifacts). Naturally, we can do whatever we want there; if we want _all_ the binaries that bumps my image (naively) to 1.5gig. I'm not sure what the difference between them is at present, but could track it down.

If it's blocking this change, I'd be happy to just delete the code around building the image in Nix and keep the source-code version for now. The main reason I wanted it is for the "external" build ability; i.e.

```
nix build github:silky/diagonal-b6/add-nix#b6-image
./result | docker load
```

That can be done anywhere, anytime (i.e. you don't need the b6 source explicitly checked out).

### Regarding the one point of "nix-awareness" that can't be escaped: _gomod2nix_.

Yes, it's true, you need to regenerated that file if you change `go.mod`. I'm sad about it too, but, for me, the benefits you get for **incremental** builds is worth the effort. Indeed, this is the main reason I added the CI job; to check that that hasn't happened.

I agree that this means, technically, this is a breaking change and forces people to care about Nix here when they might not otherwise want to. But I'm hoping that this is something we can compromise on, in order to get nice incremental builds, _and_ all the other longer-term benefits of Nix (caching builds between developers/CI with [Cachix](https://cachix.org/), perhaps [eventual adoption of b6 into nixpkgs](https://search.nixos.org/packages), etc.)

#### Slightly more detail - no implications of no gomod2nix

I was thinking about what would happen if we just dropped the gomod2nix part. It would be possible to just keep the _devShell_ parts, which would let us build the go code inside the shell only, and drop the gomod2nix parts. The problem is, [the go _binaries_ are required by the Python build](https://github.com/diagonalworks/diagonal-b6/pull/295/files#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R141), so we would _also_ lose the Python package, which is a real shame, because the Python package is what enables the very simple/nice import style that [you see here](https://github.com/silky/example-b6-analysis/blob/main/flake.nix#L19).

### Summary

Cons:

- gomod2nix needs to be run when go.mod is changed
- A few hacks end up in the flake.nix to make the build compatible between the Makefile (incremental) and the Nix derivations (the devShell, and the package entrypoints)

Pros:

- Standard Nix benefits
- Handy entrypoints for dependencies/binaries at any arbitrary commit (`nix run github:..`, `nix build github:...`).